### PR TITLE
finish the dataset preprocessing part. Just have to find a way to han…

### PR DESCRIPTION
**Pull request by eksmok**
Here i tried to create a dataset module in order to centralize the data used of the buid_tree functions (which i did not put on github yet).

### CODE REVIEW

File: src/dataset.py [hunk 1]
```diff
@@ -1,24 +1,94 @@
...
+        with open(path, newline=new_line) as csvfile:
+            data = []
+            reader = csv.reader(csvfile, delimiter=csv_limiter)
+            for row in reader:
+                data.append(row)
 
-    def chose_best_split(self):
-        # TODO : This function contains the algo which allows us to take the split from the dataset. I have to think about the link of a condition
-        pass
+            label_index = data[0].index(label_name)
```
Comment (mwolinska) [inline src/dataset.py:25]: I'd be careful here as I wasn't sure what label_index was about exactly.
As I understand you're getting the top row in order to get the names of each column? If so you are getting 2 things: all the feature names and the name of the label "category". Then within the label column you will have multiple labels. 
I would try to make that processing clear now so it's easier to use later on in your code.

E.g. label category might be "animal" and then the actual labels might be "cat", "dog"

File: src/dataset.py [hunk 2]
```diff
@@ -1,24 +1,94 @@
...
-        pass
+        dataset = Dataset(dataset=data, labels_index=label_index)
+        return dataset
+
+    @staticmethod
+    def load_from_array(data: np.array, label_index: int) -> 'Dataset':
+        dataset = Dataset(dataset=data, labels_index=label_index)
+        return dataset
+
+    def _get_possible_split(self):
```
Comment (mwolinska) [inline src/dataset.py:37]: I would keep the dataset class as simple as possible and move the split function elsewhere. All the split functions can be grouped together in a separate class that takes a dataset or even within the tree itself as you'll need to call them recursively.

File: src/dataset.py [hunk 3]
```diff
@@ -1,24 +1,94 @@
...
+        all_possibilites = {}
+
+        for i in range(self._data.shape[0]):
+            if i == self._labels_index:
```
Comment (mwolinska) [inline src/dataset.py:41]: Using a while loop seems more adequate here rather than "continue"

File: src/dataset.py [hunk 4]
```diff
@@ -1,24 +1,94 @@
...
+        dataset = Dataset(dataset=data, labels_index=label_index)
+        return dataset
+
+    @staticmethod
+    def load_from_array(data: np.array, label_index: int) -> 'Dataset':
+        dataset = Dataset(dataset=data, labels_index=label_index)
+        return dataset
+
+    def _get_possible_split(self):
+        all_possibilites = {}
```
Comment (mwolinska) [inline src/dataset.py:38]: If I understand correctly, you're getting all the possible split values at this point? If so is it necessary for the all_possibilities to be a dict? can it not be a list?

File: src/dataset.py [hunk 5]
```diff
@@ -1,24 +1,94 @@
...
+
+    def _best_split_for_each_feature(self, purity_evaluation: Information_gain):
+        ig_gain_by_feature = {}
+        possibilites = self._get_possible_split()
 
+        for i in range(self._data.shape[0]):
+            ig_gain_max = -np.inf
+            best_value = None
+            best_feature_index = None
+            if i == self._labels_index:
```
Comment (mwolinska) [inline src/dataset.py:55]: again I think a while loop would be better here

File: src/dataset.py [hunk 6]
```diff
@@ -1,24 +1,94 @@
...
+        for i in range(self._data.shape[0]):
+            if i == self._labels_index:
+                continue
+
+            all_possibilites[i] = np.unique(self._data[:, i])
+        return all_possibilites
+
+    def _best_split_for_each_feature(self, purity_evaluation: Information_gain):
+        ig_gain_by_feature = {}
+        possibilites = self._get_possible_split()
```
Comment (mwolinska) [inline src/dataset.py:49]: a bit pedantic but maybe I'd change possibilities to something like split_value_possibilities just to make it easier to follow later on

File: src/dataset.py [hunk 7]
```diff
@@ -1,24 +1,94 @@
...
+
+        for i in range(self._data.shape[0]):
+            if i == self._labels_index:
+                continue
+
+            all_possibilites[i] = np.unique(self._data[:, i])
+        return all_possibilites
+
+    def _best_split_for_each_feature(self, purity_evaluation: Information_gain):
+        ig_gain_by_feature = {}
```
Comment (mwolinska) [inline src/dataset.py:48]: again not sure if we need a dict rather than a list?

File: src/dataset.py [hunk 8]
```diff
@@ -1,24 +1,94 @@
...
+                    best_feature_index = i
+
+            ig_gain_by_feature[i] = [ig_gain_max, best_value, best_feature_index]
+        return ig_gain_by_feature
+
+    def chose_best_split(self, purity_evaluation: Information_gain):
+        ig_dict = self._best_split_for_each_feature(purity_evaluation)
+        max_ig = 0
+        split = None
+        for value_ig in ig_dict.values():
```
Comment (mwolinska) [inline src/dataset.py:73]: I find the name value_ig confusing. I thought it was the information gain value, but it seems you are storing the split value in it? i.e. split has an information gain and the split occurs at a value of the particular feature.
Would rename the overall variable to be more clear

**Review by mwolinska — APPROVED** [review]
Left a few thoughts about the dataset class. Looks like a good baseline.

I think you don't need to commit the test.py file since it's empty? In principle you should only commit changes that bring something to your code rather than placeholders.

File: src/dataset.py [hunk 9]
```diff
@@ -1,24 +1,94 @@
...
+
+    @staticmethod
+    def load_from_array(data: np.array, label_index: int) -> 'Dataset':
+        dataset = Dataset(dataset=data, labels_index=label_index)
+        return dataset
+
+    def _get_possible_split(self):
+        all_possibilites = {}
+
+        for i in range(self._data.shape[0]):
```
Comment (sachahu1) [inline src/dataset.py:40]:
```suggestion
        for row in self._data:
```
or 
```suggestion
        for row_index in range(self._data.shape[0]):
```

File: src/dataset.py (hunk 5 above)
Comment (sachahu1) [reply src/dataset.py:55]: ignore because ok

File: src/dataset.py (hunk 6 above)
Comment (sachahu1) [reply src/dataset.py:49]: Yes

Comment (sachahu1) [reply src/dataset.py:49]: Although possible_split_values is better

**Pull request merged**